File: src/cli/output-name/root-query-token.test.ts
import { describe, expect, test } from "vitest";

import type { ParsedRootQuery } from "../root-query/parsed-root-query.js";
import { rootQueryToken } from "./root-query-token.js";

describe("rootQueryToken", () => {
  test("name → identifier verbatim", () => {
    const q: ParsedRootQuery = { kind: "name", name: "value", raw: "value" };
    expect(rootQueryToken(q)).toBe("value");
  });

  test("line → l<n>", () => {
    const q: ParsedRootQuery = { kind: "line", line: 42, raw: "42" };
    expect(rootQueryToken(q)).toBe("l42");
  });

  test("line-name → l<n>-<id>", () => {
    const q: ParsedRootQuery = {
      kind: "line-name",
      line: 42,
      name: "render",
      raw: "42:render",
    };
    expect(rootQueryToken(q)).toBe("l42-render");
  });

  test("range → l<start>-<end>", () => {
    const q: ParsedRootQuery = {
      kind: "range",
      start: 10,
      end: 12,
      raw: "10-12",
    };
    expect(rootQueryToken(q)).toBe("l10-12");
  });

  test("range-name → l<start>-<end>-<id>", () => {
    const q: ParsedRootQuery = {
      kind: "range-name",
      start: 10,
      end: 12,
      name: "render",
      raw: "10-12:render",
    };
    expect(rootQueryToken(q)).toBe("l10-12-render");
  });
});

File: crates/unsnarl/src/cli/run_cli/resolve_output_path/root_query_token_test.rs
use super::*;

use unsnarl_ir::SourceLine;

#[test]
fn name_uses_identifier_verbatim() {
    let q = ParsedRootQuery::Name {
        name: "value".to_string(),
        raw: "value".to_string(),
    };
    assert_eq!(root_query_token(&q), "value");
}

#[test]
fn line_yields_l_n() {
    let q = ParsedRootQuery::Line {
        line: SourceLine(42),
        raw: "42".to_string(),
    };
    assert_eq!(root_query_token(&q), "l42");
}

#[test]
fn line_name_yields_l_n_dash_id() {
    let q = ParsedRootQuery::LineName {
        line: SourceLine(42),
        name: "render".to_string(),
        raw: "42:render".to_string(),
    };
    assert_eq!(root_query_token(&q), "l42-render");
}

#[test]
fn range_yields_l_start_dash_end() {
    let q = ParsedRootQuery::Range {
        start: SourceLine(10),
        end: SourceLine(12),
        raw: "10-12".to_string(),
    };
    assert_eq!(root_query_token(&q), "l10-12");
}

#[test]
fn range_name_yields_l_start_dash_end_dash_id() {
    let q = ParsedRootQuery::RangeName {
        start: SourceLine(10),
        end: SourceLine(12),
        name: "render".to_string(),
        raw: "10-12:render".to_string(),
    };
    assert_eq!(root_query_token(&q), "l10-12-render");
}

#[test]
fn line_or_name_normalizes_to_plain_line_shape() {
    let q = ParsedRootQuery::LineOrName {
        line: SourceLine(12),
        name: "L12".to_string(),
        raw: "L12".to_string(),
    };
    assert_eq!(root_query_token(&q), "l12");
}
